playground/src/App.tsx

```diff
@@ -0,0 +1,200 @@
+  config,
+  groupName,
+  fieldName,
+  onOptionChange,
+}: {
+  config: Config | null;
+  groupName: string;
+  fieldName: string;
+  onOptionChange: (groupName: string, fieldName: string, value: string) => void;
+}) {
+  const value =
+    config && config[groupName] && config[groupName][fieldName]
+      ? config[groupName][fieldName]
+      : "";
+
+  return (
+    <span>
+      <label>
+        {fieldName}
+        <input
+          value={value}
+          placeholder={defaultConfig[groupName][fieldName]}
+          type="text"
+          onChange={(event) => {
+            onOptionChange(groupName, fieldName, event.target.value);
+          }}
+        />
+      </label>
+    </span>
+  );
+}
+
+function Options({
+  config,
+  onOptionChange,
+}: {
+  config: Config | null;
+  onOptionChange: (groupName: string, fieldName: string, value: string) => void;
+}) {
+  return (
+    <div className="options">
+      {AVAILABLE_OPTIONS.map((group) => (
+        <details key={group.name}>
+          <summary>{group.name}</summary>
+          <div>
+            <ul>
+              {group.fields.map((field) => (
+                <li key={field.name}>
+                  <OptionEntry
+                    config={config}
+                    groupName={group.name}
+                    fieldName={field.name}
+                    onOptionChange={onOptionChange}
+                  />
+                </li>
+              ))}
+            </ul>
+          </div>
+        </details>
+      ))}
+    </div>
+  );
+}
+
+function App() {
+  const monaco = useMonaco();
+  const [ruffInitialized, setRuffInitialized] = useState<boolean>(false);
+  const [config, setConfig] = useState<Config | null>(null);
+  const [source, setSource] = useState<string | null>(null);
+  const [error, setError] = useState<string | null>(null);
+
+  useEffect(() => {
+    init().then(() => setRuffInitialized(true));
+  }, []);
+
+  useEffect(() => {
+    if (source === null && config === null && monaco) {
+      let [config, source] = restoreConfigAndSource();
+      setConfig(config);
+      setSource(source);
+    }
+  }, [monaco]);
+
+  useEffect(() => {
+    if (config != null && source != null) {
+      persistConfigAndSource(config, source);
+    }
+  }, [config, source]);
+
+  useEffect(() => {
+    let editor = monaco?.editor;
+    let model = editor?.getModels()[0];
+    if (
+      !editor ||
+      !model ||
+      !ruffInitialized ||
+      source === null ||
+      config === null
+    ) {
+      return;
+    }
+
+    let checks: Check[];
+    try {
+      checks = check(source, toRuffConfig(config));
+      setError(null);
+    } catch (e) {
+      setError(String(e));
+      return;
+    }
+
+    editor.setModelMarkers(
+      model,
+      "owner",
+      checks.map((check) => ({
+        startLineNumber: check.location.row,
+        startColumn: check.location.column,
+        endLineNumber: check.end_location.row,
+        endColumn: check.end_location.column,
```

Comment:
Any idea why these aren't spanning the entire range?

<img width="119" alt="Screen Shot 2022-12-23 at 11 23 55 PM" src="https://user-images.githubusercontent.com/1309177/209421068-cd2b5e92-6e74-4bf8-9a48-0902a0720b1f.png">

Reply:
Classic off-by-one error. Column indices needed to be adjusted for the 1-based index of the editor.

[Works now](https://611ef32d.ruff-axw.pages.dev/#N4XwJBCWC2AOD2AnALgAngZwHQHcAWAhsgKYBuxiQA)